### PR TITLE
Vault-17911: add support for hex values in oid extension

### PR DESCRIPTION
**Pull request by rculpepper**

### CODE REVIEW

File: builtin/credential/cert/path_login.go [hunk 1]
```diff
@@ -507,17 +508,29 @@ func (b *backend) matchesCertificateExtensions(clientCert *x509.Certificate, con
...
 		var parsedValue string
 		asn1.Unmarshal(ext.Value, &parsedValue)
 		clientExtMap[ext.Id.String()] = parsedValue
+		hexExtMap[ext.Id.String()] = hex.EncodeToString(ext.Value)
 	}
-	// If any of the required extensions don'log match the constraint fails
+
+	// If any of the required extensions don't match the constraint fails
 	for _, requiredExt := range config.Entry.RequiredExtensions {
 		reqExt := strings.SplitN(requiredExt, ":", 2)
```
Comment (stevendpclark) [inline builtin/credential/cert/path_login.go:522]: I believe we have an existing bug here in which we don't validate we split the expected number of parts within the `requiredExt`,

File: builtin/credential/cert/path_login.go [hunk 2]
```diff
@@ -507,17 +508,29 @@ func (b *backend) matchesCertificateExtensions(clientCert *x509.Certificate, con
 	// including its ASN.1 type tag bytes. For the sake of simplicity, assume string type
 	// and drop the tag bytes. And get the number of bytes from the tag.
 	clientExtMap := make(map[string]string, len(clientCert.Extensions))
+	hexExtMap := make(map[string]string, len(clientCert.Extensions))
+
 	for _, ext := range clientCert.Extensions {
 		var parsedValue string
 		asn1.Unmarshal(ext.Value, &parsedValue)
```
Comment (stevendpclark) [inline builtin/credential/cert/path_login.go:515]: Kinda odd we ignore the unmarshalling error here?

Comment (cipherboy) [reply builtin/credential/cert/path_login.go:515]: Yes, but this is existing behavior; it results in `map[<oid>] = ""` in the map, which you _could_ match on in the API by providing just the OID.

File: builtin/credential/cert/path_login.go (hunk 1 above)
Comment (cipherboy) [reply builtin/credential/cert/path_login.go:522]: Interestingly, we do have:

```
			testAccStepCert(t, "web", ca, "foo", allowed{ext: "2.1.1.1:"}, false),
			testAccStepLoginInvalid(t, connState),
			testAccStepCert(t, "web", ca, "foo", allowed{ext: "2.1.1.1:,2.1.1.2:*"}, false),
			testAccStepLoginInvalid(t, connState),```

as test cases, which seems to suggest this should be rejected for login, but not from the API, which is weird. I'm not entirely sure why this wouldn't work though, tbh.

File: builtin/credential/cert/path_login.go (hunk 2 above)
Comment (stevendpclark) [reply builtin/credential/cert/path_login.go:515]: Could we codify it explicitly so I don't break existing customers, in the future thinking I'm fixing a bug :D

Comment (cipherboy) [reply builtin/credential/cert/path_login.go:515]: Interestingly, we do have:

```
			testAccStepCert(t, "web", ca, "foo", allowed{ext: "2.1.1.1:"}, false),
			testAccStepLoginInvalid(t, connState),
			testAccStepCert(t, "web", ca, "foo", allowed{ext: "2.1.1.1:,2.1.1.2:*"}, false),
			testAccStepLoginInvalid(t, connState),
```

as test cases, which seems to suggest this should be rejected for login, but not from the API, which is weird. I'm not entirely sure why this wouldn't work though, tbh.

Comment (cipherboy) [reply builtin/credential/cert/path_login.go:515]: Ah, but these are string extensions. You'd need to do this:

```
testAccStepCert(t, "web", ca, "foo", allowed{ext: "2.5.29.17:"}, false),
testAccStepLogin(t, connState),
```

which seems to succeed.

File: builtin/credential/cert/path_login.go (hunk 1 above)
Comment (cipherboy) [reply builtin/credential/cert/path_login.go:522]: This panics; @rculpepper I'd add a check for `len(reqExt) == 2` here before continuing and return a different error if not (saying something like `malformed required_extension on cert auth role`).

File: builtin/credential/cert/path_login.go (hunk 2 above)
Comment (cipherboy) [reply builtin/credential/cert/path_login.go:515]: @rculpepper I'd probably add this as an explicit test case btw. :-)

Comment (stevendpclark) [reply builtin/credential/cert/path_login.go:515]: Okay, if we can't change this behavior, or it is what we want, could we explicitly, trap the error and set the value to "" so the code is explicit about the behavior we want.

**Review by cipherboy — APPROVED** [review]

**Review by stevendpclark — APPROVED** [review]

**Comment by cipherboy** [comment]
Thank you @rculpepper!

**Pull request merged**